File: chat2db-community-server/chat2db-community-domain/chat2db-community-domain-api/src/main/java/ai/chat2db/community/domain/api/model/task/CsvOptions.java
package ai.chat2db.community.domain.api.model.task;

import ai.chat2db.community.tools.exception.BusinessException;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.apache.commons.lang3.StringUtils;

import java.nio.charset.Charset;
import java.util.LinkedHashMap;
import java.util.Locale;
import java.util.Map;
import java.util.Set;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class CsvOptions {

    public static final String DEFAULT_ENCODING = "UTF-8";
    public static final String AUTO_ENCODING = "AUTO";
    public static final String DEFAULT_DELIMITER = ",";
    public static final String DEFAULT_QUOTE = "\"";
    public static final String DEFAULT_ESCAPE = "\"";
    public static final String DEFAULT_NEWLINE = "LF";
    public static final String DEFAULT_DATE_ORDER = "YMD";
    public static final String DEFAULT_DATE_TIME_ORDER = "DATE_TIME";
    public static final String DEFAULT_DATE_DELIMITER = "-";
    public static final String DEFAULT_TIME_DELIMITER = ":";
    public static final String DEFAULT_DECIMAL_SYMBOL = ".";

    private static final Set<String> SUPPORTED_NEWLINES = Set.of("LF", "CRLF", "CR");
    private static final Set<String> SUPPORTED_DATE_ORDERS = Set.of("YMD", "YDM", "MDY", "MYD", "DMY", "DYM");
    private static final Set<String> SUPPORTED_DATE_TIME_ORDERS = Set.of(
            "DATE_TIME", "TIME_DATE", "DATE_TIME_TIMEZONE", "TIME_DATE_TIMEZONE", "TIME_TIMEZONE_DATE");

    private String encoding;

    private String delimiter;

    private String quote;

    private String escape;

    private String newline;

    private Boolean hasHeader;

    private Boolean emptyAsNull;

    private Integer headerRow;

    private Integer dataStartRow;

    private Integer dataEndRow;

    private String dateOrder;

    private String dateTimeOrder;

    private String dateDelimiter;

    private String yearDelimiter;

    private String timeDelimiter;

    private String decimalSymbol;

    public static CsvOptions defaults() {
        return CsvOptions.builder()
                .encoding(DEFAULT_ENCODING)
                .delimiter(DEFAULT_DELIMITER)
                .quote(DEFAULT_QUOTE)
                .escape(DEFAULT_ESCAPE)
                .newline(DEFAULT_NEWLINE)
                .hasHeader(true)
                .emptyAsNull(true)
                .headerRow(1)
                .dataStartRow(2)
                .dateOrder(DEFAULT_DATE_ORDER)
                .dateTimeOrder(DEFAULT_DATE_TIME_ORDER)
                .dateDelimiter(DEFAULT_DATE_DELIMITER)
                .yearDelimiter(DEFAULT_DATE_DELIMITER)
                .timeDelimiter(DEFAULT_TIME_DELIMITER)
                .decimalSymbol(DEFAULT_DECIMAL_SYMBOL)
                .build();
    }

    public static CsvOptions fromMap(Map<String, Object> values) {
        CsvOptions defaults = defaults();
        if (values == null || values.isEmpty()) {
            return defaults.validate();
        }
        CsvOptions options = CsvOptions.builder()
                .encoding(stringValue(values.get("encoding"), defaults.getEncoding()))
                .delimiter(stringValue(values.get("delimiter"), defaults.getDelimiter()))
                .quote(stringValue(values.get("quote"), defaults.getQuote()))
                .escape(stringValue(values.get("escape"), defaults.getEscape()))
                .newline(stringValue(values.get("newline"), defaults.getNewline()))
                .hasHeader(booleanValue(values.get("hasHeader"), defaults.getHasHeader()))
                .emptyAsNull(booleanValue(values.get("emptyAsNull"), defaults.getEmptyAsNull()))
                .headerRow(integerValue(values.get("headerRow")))
                .dataStartRow(integerValue(values.get("dataStartRow")))
                .dataEndRow(integerValue(values.get("dataEndRow")))
                .dateOrder(stringValue(values.get("dateOrder"), defaults.getDateOrder()))
                .dateTimeOrder(stringValue(values.get("dateTimeOrder"), defaults.getDateTimeOrder()))
                .dateDelimiter(stringValue(values.get("dateDelimiter"), defaults.getDateDelimiter()))
                .yearDelimiter(stringValue(values.get("yearDelimiter"), defaults.getYearDelimiter()))
                .timeDelimiter(stringValue(values.get("timeDelimiter"), defaults.getTimeDelimiter()))
                .decimalSymbol(stringValue(values.get("decimalSymbol"), defaults.getDecimalSymbol()))
                .build();
        return options.validate();
    }

    public CsvOptions validate() {
        CsvOptions options;
        try {
            options = normalized();
        } catch (Exception e) {
            throw new BusinessException("import.preview.invalidEncoding", new Object[]{encoding}, e);
        }
        if (!isSingleTextCharacter(options.delimiter)
                || !SUPPORTED_NEWLINES.contains(options.newline)
                || !isSingleTextCharacter(options.quote)
                || !isSingleTextCharacter(options.escape)
                || options.delimiter.equals(options.quote)
                || options.delimiter.equals(options.escape)
                || !SUPPORTED_DATE_ORDERS.contains(options.dateOrder)
                || !SUPPORTED_DATE_TIME_ORDERS.contains(options.dateTimeOrder)
                || !isSingleTextCharacter(options.dateDelimiter)
                || !isSingleTextCharacter(options.yearDelimiter)
                || !isSingleTextCharacter(options.timeDelimiter)
                || !(".".equals(options.decimalSymbol) || ",".equals(options.decimalSymbol))
                || options.headerRow < 1 || options.dataStartRow < 1
                || options.dataEndRow != null && options.dataEndRow < options.dataStartRow
                || Boolean.TRUE.equals(options.hasHeader) && options.headerRow >= options.dataStartRow) {
            throw new BusinessException("import.preview.invalidCsvOptions");
        }
        if (!AUTO_ENCODING.equals(options.encoding)) {
            try {
                Charset.forName(options.encoding);
            } catch (Exception e) {
                throw new BusinessException("import.preview.invalidEncoding", new Object[]{options.encoding}, e);
            }
        }
        return options;
    }

    public Map<String, Object> toMap() {
        CsvOptions options = validate();
        Map<String, Object> values = new LinkedHashMap<>();
        values.put("encoding", options.encoding);
        values.put("delimiter", options.delimiter);
        values.put("quote", options.quote);
        values.put("escape", options.escape);
        values.put("newline", options.newline);
        values.put("hasHeader", options.hasHeader);
        values.put("emptyAsNull", options.emptyAsNull);
        values.put("headerRow", options.headerRow);
        values.put("dataStartRow", options.dataStartRow);
        values.put("dataEndRow", options.dataEndRow);
        values.put("dateOrder", options.dateOrder);
        values.put("dateTimeOrder", options.dateTimeOrder);
        values.put("dateDelimiter", options.dateDelimiter);
        values.put("yearDelimiter", options.yearDelimiter);
        values.put("timeDelimiter", options.timeDelimiter);
        values.put("decimalSymbol", options.decimalSymbol);
        return values;
    }

    public String rowSeparator() {
        return switch (validate().newline) {
            case "CRLF" -> "\r\n";
            case "CR" -> "\r";
            default -> "\n";
        };
    }

    private CsvOptions normalized() {
        CsvOptions defaults = defaults();
        boolean normalizedHasHeader = hasHeader == null ? defaults.hasHeader : hasHeader;
        int normalizedHeaderRow = headerRow == null ? defaults.headerRow : headerRow;
        int normalizedDataStartRow = dataStartRow == null
                ? (normalizedHasHeader ? normalizedHeaderRow + 1 : 1) : dataStartRow;
        return CsvOptions.builder()
                .encoding(normalizeEncoding(StringUtils.defaultIfBlank(encoding, defaults.encoding)))
                .delimiter(StringUtils.defaultIfEmpty(delimiter, defaults.delimiter))
                .quote(StringUtils.defaultIfEmpty(quote, defaults.quote))
                .escape(StringUtils.defaultIfEmpty(escape, defaults.escape))
                .newline(StringUtils.defaultIfBlank(newline, defaults.newline).trim().toUpperCase(Locale.ROOT))
                .hasHeader(normalizedHasHeader)
                .emptyAsNull(emptyAsNull == null ? defaults.emptyAsNull : emptyAsNull)
                .headerRow(normalizedHeaderRow)
                .dataStartRow(normalizedDataStartRow)
                .dataEndRow(dataEndRow)
                .dateOrder(normalizeOption(dateOrder, defaults.dateOrder))
                .dateTimeOrder(normalizeOption(dateTimeOrder, defaults.dateTimeOrder))
                .dateDelimiter(StringUtils.defaultIfEmpty(dateDelimiter, defaults.dateDelimiter))
                .yearDelimiter(StringUtils.defaultIfEmpty(yearDelimiter,
                        StringUtils.defaultIfEmpty(dateDelimiter, defaults.dateDelimiter)))
                .timeDelimiter(StringUtils.defaultIfEmpty(timeDelimiter, defaults.timeDelimiter))
                .decimalSymbol(StringUtils.defaultIfEmpty(decimalSymbol, defaults.decimalSymbol))
                .build();
    }

    private static String normalizeOption(String value, String defaultValue) {
        return StringUtils.defaultIfBlank(value, defaultValue).trim().toUpperCase(Locale.ROOT);
    }

    private static String normalizeEncoding(String value) {
        String normalized = value.trim().toUpperCase(Locale.ROOT);
        if (AUTO_ENCODING.equals(normalized)) {
            return AUTO_ENCODING;
        }
        return Charset.forName(value.trim()).name().toUpperCase(Locale.ROOT);
    }

    private static boolean isSingleTextCharacter(String value) {
        return value != null && value.length() == 1 && value.charAt(0) != '\n' && value.charAt(0) != '\r';
    }

    private static String stringValue(Object value, String defaultValue) {
        if (value == null) {
            return defaultValue;
        }
        if (value instanceof String stringValue) {
            return stringValue;
        }
        throw new BusinessException("import.preview.invalidCsvOptions");
    }

    private static Boolean booleanValue(Object value, Boolean defaultValue) {
        if (value == null) {
            return defaultValue;
        }
        if (value instanceof Boolean booleanValue) {
            return booleanValue;
        }
        throw new BusinessException("import.preview.invalidCsvOptions");
    }

    private static Integer integerValue(Object value) {
        if (value == null) {
            return null;
        }
        if (value instanceof Number number) {
            long longValue = number.longValue();
            if (longValue < Integer.MIN_VALUE || longValue > Integer.MAX_VALUE
                    || number.doubleValue() != longValue) {
                throw new BusinessException("import.preview.invalidCsvOptions");
            }
            return (int) longValue;
        }
        throw new BusinessException("import.preview.invalidCsvOptions");
    }
}

File: chat2db-community-client/src/blocks/ImportAndExport/utils/csvOptions.ts
import type { ICsvOptions } from '@/typings/importExport';
import { ImportExportFileType, ImportPreviewErrorCode } from '@/constants/importExport';

export const DEFAULT_CSV_OPTIONS: ICsvOptions = {
  encoding: 'AUTO',
  delimiter: ',',
  quote: '"',
  escape: '"',
  newline: 'LF',
  hasHeader: true,
  emptyAsNull: true,
  headerRow: 1,
  dataStartRow: 2,
  dataEndRow: undefined,
  dateOrder: 'YMD',
  dateTimeOrder: 'DATE_TIME',
  dateDelimiter: '-',
  yearDelimiter: '-',
  timeDelimiter: ':',
  decimalSymbol: '.',
};

export class CsvOptionsValidationError extends Error {
  readonly errorCode = ImportPreviewErrorCode.INVALID_CSV_OPTIONS;

  constructor() {
    super();
    this.name = 'CsvOptionsValidationError';
  }
}

const SUPPORTED_NEWLINES = ['LF', 'CRLF', 'CR'];
const SUPPORTED_DATE_ORDERS = ['YMD', 'YDM', 'MDY', 'MYD', 'DMY', 'DYM'];
const SUPPORTED_DATE_TIME_ORDERS = [
  'DATE_TIME',
  'TIME_DATE',
  'DATE_TIME_TIMEZONE',
  'TIME_DATE_TIMEZONE',
  'TIME_TIMEZONE_DATE',
];

export function validateCsvOptions(options: ICsvOptions): ICsvOptions {
  const normalized: ICsvOptions = {
    encoding: (options.encoding || DEFAULT_CSV_OPTIONS.encoding).trim().toUpperCase(),
    delimiter: options.delimiter || DEFAULT_CSV_OPTIONS.delimiter,
    quote: options.quote || DEFAULT_CSV_OPTIONS.quote,
    escape: options.escape || DEFAULT_CSV_OPTIONS.escape,
    newline: (options.newline || DEFAULT_CSV_OPTIONS.newline).trim().toUpperCase(),
    hasHeader: options.hasHeader ?? DEFAULT_CSV_OPTIONS.hasHeader,
    emptyAsNull: options.emptyAsNull ?? DEFAULT_CSV_OPTIONS.emptyAsNull,
    headerRow: options.headerRow ?? DEFAULT_CSV_OPTIONS.headerRow,
    dataStartRow: options.dataStartRow ?? (options.hasHeader ? (options.headerRow || 1) + 1 : 1),
    dataEndRow: options.dataEndRow,
    dateOrder: options.dateOrder || DEFAULT_CSV_OPTIONS.dateOrder,
    dateTimeOrder: options.dateTimeOrder || DEFAULT_CSV_OPTIONS.dateTimeOrder,
    dateDelimiter: options.dateDelimiter || DEFAULT_CSV_OPTIONS.dateDelimiter,
    yearDelimiter: options.yearDelimiter || options.dateDelimiter || DEFAULT_CSV_OPTIONS.yearDelimiter,
    timeDelimiter: options.timeDelimiter || DEFAULT_CSV_OPTIONS.timeDelimiter,
    decimalSymbol: options.decimalSymbol || DEFAULT_CSV_OPTIONS.decimalSymbol,
  };
  if (
    !normalized.encoding ||
    normalized.delimiter.length !== 1 ||
    !SUPPORTED_NEWLINES.includes(normalized.newline) ||
    normalized.quote.length !== 1 ||
    normalized.escape.length !== 1 ||
    normalized.quote === '\n' ||
    normalized.quote === '\r' ||
    normalized.escape === '\n' ||
    normalized.escape === '\r' ||
    normalized.delimiter === normalized.quote ||
    normalized.delimiter === normalized.escape ||
    !Number.isInteger(normalized.headerRow) ||
    normalized.headerRow < 1 ||
    !Number.isInteger(normalized.dataStartRow) ||
    normalized.dataStartRow < 1 ||
    (normalized.dataEndRow !== undefined &&
      (!Number.isInteger(normalized.dataEndRow) || normalized.dataEndRow < normalized.dataStartRow)) ||
    (normalized.hasHeader && normalized.headerRow >= normalized.dataStartRow) ||
    !SUPPORTED_DATE_ORDERS.includes(normalized.dateOrder) ||
    !SUPPORTED_DATE_TIME_ORDERS.includes(normalized.dateTimeOrder) ||
    normalized.dateDelimiter.length !== 1 ||
    normalized.yearDelimiter.length !== 1 ||
    normalized.timeDelimiter.length !== 1 ||
    !['.', ','].includes(normalized.decimalSymbol)
  ) {
    throw new CsvOptionsValidationError();
  }
  return normalized;
}

export function getCsvDateTimeExamples(options: ICsvOptions): string[] {
  const values = [
    { year: '23', month: '8', day: '24' },
    { year: '2023', month: '8', day: '24' },
    { year: '23', month: 'Aug', day: '24' },
    { year: '23', month: 'August', day: '24' },
  ];
  return values.map(({ year, month, day }) => {
    const parts = { Y: year, M: month, D: day };
    const order = options.dateOrder.split('') as Array<keyof typeof parts>;
    const separators = [0, 1].map((index) => {
      const nextToYear = order[index] === 'Y' || order[index + 1] === 'Y';
      return nextToYear ? options.yearDelimiter : options.dateDelimiter;
    });
    const date = `${parts[order[0]]}${separators[0]}${parts[order[1]]}${separators[1]}${parts[order[2]]}`;
    const time = `15${options.timeDelimiter}30${options.timeDelimiter}38`;
    switch (options.dateTimeOrder) {
      case 'TIME_DATE':
        return `${time} ${date}`;
      case 'DATE_TIME_TIMEZONE':
        return `${date} ${time} +08:00`;
      case 'TIME_DATE_TIMEZONE':
        return `${time} ${date} +08:00`;
      case 'TIME_TIMEZONE_DATE':
        return `${time} +08:00 ${date}`;
      default:
        return `${date} ${time}`;
    }
  });
}

export function buildCsvOptionsForTaskSubmit(isCsv: boolean, options: ICsvOptions): ICsvOptions | undefined {
  return isCsv ? validateCsvOptions(options) : undefined;
}

export function csvOptionsToPreviewParam(isCsv: boolean, options: ICsvOptions): string | undefined {
  const validated = buildCsvOptionsForTaskSubmit(isCsv, options);
  return validated ? JSON.stringify(validated) : undefined;
}

export function inferImportFileFormat(filePath: string): ImportExportFileType {
  const lower = filePath.toLowerCase();
  if (lower.endsWith('.xlsx')) {
    return ImportExportFileType.XLSX;
  }
  if (lower.endsWith('.xls')) {
    return ImportExportFileType.XLS;
  }
  if (lower.endsWith('.json')) {
    return ImportExportFileType.JSON;
  }
  if (lower.endsWith('.sql')) {
    return ImportExportFileType.SQL;
  }
  return ImportExportFileType.CSV;
}

export function supportsCsvMappingPreview(format?: ImportExportFileType): boolean {
  return format === ImportExportFileType.CSV;
}
